### PR TITLE
[Track2 Function#4] Add abstract implementation for function library

### PR DESCRIPTION
**Pull request by Flanker32**
Add abstract implementation for function library

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java [hunk 1]
```diff
@@ -45,6 +48,35 @@ private AzureAppService(@Nonnull final List<Subscription> subscriptions) {
...
+    }
+
+    public List<IFunctionApp> functionApps() {
+        return getSubscriptions().stream()
+                .map(subscription -> getAzureResourceManager(subscription.getId()))
+                .flatMap(azureResourceManager -> azureResourceManager.functionApps().list().stream())
+                .collect(Collectors.toList()).stream()
+                .filter(webAppBasic -> StringUtils.containsIgnoreCase(webAppBasic.innerModel().kind(), "functionapp")) // Filter out function apps
+                .map(webAppBasic -> functionApp(webAppBasic.id()))
+                .collect(Collectors.toList());
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:77]: twice colletors.toList()

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/FunctionAppDeploymentSlot.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+
+public class FunctionAppDeploymentSlot extends AbstractAppService<FunctionDeploymentSlot, FunctionAppDeploymentSlotEntity>
+        implements IFunctionAppDeploymentSlot {
+
+    private final AzureResourceManager azureClient;
+
+    public FunctionAppDeploymentSlot(FunctionAppDeploymentSlotEntity deploymentSlot, AzureResourceManager azureClient) {
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/FunctionAppDeploymentSlot.java:26]: add refer FunctionApp property here.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/FunctionAppDeploymentSlot.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    @Nullable
+    @Override
+    protected FunctionDeploymentSlot remote() {
+        final FunctionApp parentFunctionApp = getParentWebApp();
+        return StringUtils.isNotEmpty(entity.getId()) ?
+                parentFunctionApp.deploymentSlots().getById(entity.getId()) :
+                parentFunctionApp.deploymentSlots().getByName(entity.getName());
+    }
+
+    private FunctionApp getParentWebApp() {
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/FunctionAppDeploymentSlot.java:76]: typo,... WebApp

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java (hunk 1 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:77]: Will address in later pr

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/FunctionAppDeploymentSlot.java (hunk 3 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/FunctionAppDeploymentSlot.java:76]: Will address in later pr

**Review by shenqianjin — APPROVED** [review]

**Pull request merged**